### DISCUSSION TITLE
Changing graph layout and vertices

### DISCUSSION THREAD

**Discussion by austincbaker** [NationalSecurityAgency/ghidra#4441] [Q&A]
I've been trying to figure out how to set a specific layout for my nodes in a plugin I'm working on. The issue I keep running into is that first, my nodes location is always set to null and I can'f figure out a way to edit them that works with the layout provider, or how to just set them using one of the Jung layouts. I also was wondering how to set a specific color/ shape for specific nodes. For example, if I wanted the initial node to be a green circle and the terminal node to be a red triangle.
My nodes extend DockingVisualVertex
my edges extend AbstractVisualEdge
and my graph extends FilteringVisualGraph
if those help any.

**Comment by dragonmacher** [comment]
Hello and welcome to the aspiring graph API.    Most of the above classes are my creation and are likely in a state of stale nascency, waiting to be fully developed.    Some of the classes you mentioned above were created as a simplified entry point into the API.    Further, the `DockingVisualVertex` is currently only used in our system as part of a demo graph that was meant to be used as an example or reference implementation for using the API.

You should be able to use the `SampleGraphProvider` as a guide for how to use the API.  For example, regarding your layout question, this class offers multiple layouts for the client to use.  You should be able to statically see how it is wired.   Further, if you wish, you can load this plugin into Ghidra to see what it produces and to debug the various features as they are used.

The basic method used by the Sample Graph Plugin and the other graphs in the system is to offload layout position calculation to a background thread.   The background thread will use the active layout to calculate vertex and edge positions via a call to `VisualGraphLayout.calculateLocations(VisualGraph<V, E> graph, TaskMonitor monitor)`.    When complete, the task that called that method will take the resulting layout positions and update the layout.  This updating is currently done using an animator class to enable transitioning from the old positions to the new positions.

If you really wanted to manipulate the vertex positions manually without using the above framework, then you should be able to do that via `VisualGraphLayout.setLocation(V v, Point2D location, ChangeType changeType)`.

**Reply by austincbaker** [reply]
I appreciate the help!

**Comment by dragonmacher** [comment]
The question regarding changing colors and shapes has a more nuanced answer.   

The basic vertex rendering in this system is done in a 'simplified' way that is meant to emulate Java's 'renderer' concept.  The graph painting will ask each vertex for it's component and then paint that component into the UI.  The benefit of this is that you can make very complicated vertices in your graph;  the downside of this approach is that simple things, such as painting shapes and colors, have become more complicated.   

I do not think we have a simple example that shows how to use this API to create simple shape/color vertex painting.  That being said, it is easily doable if you have some basic image painting knowledge.   There is a test vertex that will create custom shapes and colors for use in painting.  That can be seen here: https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Framework/Graph/src/test/java/ghidra/graph/algo/viewer/AlgorithmTestSteppingVertex.java

You can see `AlgorithmTestSteppingVertex.buildShapes()` to see how the shapes are created and turned into images.   Those images will later be used in the call to get the component for painting, as can be seen here:
```
public JComponent getComponent() {
		ShapeImage si = getShapeImage();
		ImageIcon icon = new ImageIcon(si.getImage());
		tempLabel.setIcon(icon);
		return tempLabel;
	}
```

To emulate this behavior, you can override `DockingVisualVertex.getComponent()` to do something similar to the code above.
